File: src/main/java/org/gecko/view/views/viewelement/BlockViewElement.java
package org.gecko.view.views.viewelement;

import javafx.beans.property.SimpleListProperty;
import javafx.collections.FXCollections;
import javafx.geometry.Point2D;
import javafx.scene.layout.Pane;
import lombok.Getter;
import lombok.Setter;
import org.gecko.model.Element;
import org.gecko.viewmodel.PositionableViewModelElement;

/**
 * An abstract representation of a {@link Pane} view element, that is an element with a rectangular shape in a Gecko
 * project. Contains a list of {@link Point2D edge point}s.
 */
@Getter
public abstract class BlockViewElement extends Pane {
    protected static final int BACKGROUND_ROUNDING = 15;

    private final SimpleListProperty<Point2D> edgePoints = new SimpleListProperty<>(FXCollections.observableArrayList());

    @Setter
    private boolean selected;


    protected BlockViewElement(PositionableViewModelElement<? extends Element> positionableViewModelElement) {
        // Initialize edge points for a rectangular shaped block
        for (int i = 0; i < 4; i++) {
            edgePoints.add(Point2D.ZERO);
        }

        // Auto calculate new edge points on size and position changes
        positionableViewModelElement.getSizeProperty().addListener((observable, oldValue, newValue) -> calculateEdgePoints(positionableViewModelElement));
        positionableViewModelElement.getPositionProperty().addListener((observable, oldValue, newValue) -> calculateEdgePoints(positionableViewModelElement));

        calculateEdgePoints(positionableViewModelElement);
    }

    private void calculateEdgePoints(PositionableViewModelElement<?> target) {
        Point2D position = target.getPosition();
        double width = target.getSize().getX();
        double height = target.getSize().getY();
        edgePoints.setAll(
                position.add(Point2D.ZERO),
                position.add(new Point2D(width, 0)),
                position.add(new Point2D(width, height)),
                position.add(new Point2D(0, height)));
    }
}

File: src/main/kotlin/org/gecko/view/views/viewelement/BlockViewElement.kt
package org.gecko.view.views.viewelement

import javafx.geometry.Point2D
import javafx.scene.layout.Pane
import org.gecko.viewmodel.PositionableElement
import org.gecko.viewmodel.listProperty

/**
 * An abstract representation of a [Pane] view element, that is an element with a rectangular shape
 * in a Gecko project. Contains a list of [edge point][Point2D]s.
 */
abstract class BlockViewElement(positionableElement: PositionableElement) : Pane() {
    val edgePoints = listProperty<Point2D>()

    init {
        // Initialize edge points for a rectangular shaped block
        for (i in 0..3) {
            edgePoints.add(Point2D.ZERO)
        }

        // Auto calculate new edge points on size and position changes
        positionableElement.sizeProperty.addListener { _, _, _ ->
            calculateEdgePoints(positionableElement)
        }
        positionableElement.positionProperty.addListener { _, _, _ ->
            calculateEdgePoints(positionableElement)
        }

        calculateEdgePoints(positionableElement)
    }

    fun calculateEdgePoints(target: PositionableElement) {
        val position = target.position
        val width = target.size.x
        val height = target.size.y
        edgePoints.setAll(
            position.add(Point2D.ZERO),
            position.add(Point2D(width, 0.0)),
            position.add(Point2D(width, height)),
            position.add(Point2D(0.0, height))
        )
    }
}

const val BACKGROUND_ROUNDING: Int = 15
